discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,376 @@
+            current_end = actual_end_block
+
+            last_scan_duration = int(time.time() - start)
+            all_processed += new_entries
+
+            # Print progress bar
+            if progress_callback:
+                progress_callback(
+                    current_block,
+                    end_block_timestamp,
+                    chunk_size,
+                    len(new_entries),
+                )
+
+            # Try to guess how many blocks to fetch over `eth_get_logs` API next time
+            chunk_size = self.estimate_next_chunk_size(chunk_size, len(new_entries))
+
+            # Set where the next chunk starts
+            current_block = current_end + 1
+            total_chunks_scanned += 1
+            self.state.end_chunk(min(current_end, self.get_suggested_scan_end_block()))
+
+        return all_processed, total_chunks_scanned
+
+
+def _retry_web3_call(
+    func,
+    start_block,
+    end_block,
+    retries=MAX_REQUEST_RETRIES,
+    delay=REQUEST_RETRY_SECONDS,
+) -> Union[Tuple[int, list], None]:
+    """A custom retry loop to throttle down block range.
+
+    If our JSON-RPC server cannot serve all incoming `eth_get_logs` in a single request,
+    we retry and throttle down block range for every retry.
+
+    For example, Go Ethereum does not indicate what is an acceptable response size.
+    It just fails on the server-side with a "context was cancelled" warning.
+
+    :param func: A callable that triggers Ethereum JSON-RPC, as func(start_block, end_block)
+    :param start_block: The initial start block of the block range
+    :param end_block: The initial start block of the block range
+    :param retries: How many times we retry
+    :param delay: Time to sleep between retries
+    """
+    for i in range(retries):
+        try:
+            return end_block, func(start_block, end_block)
+        except Exception as e:
+            # Assume this is HTTPConnectionPool(host='localhost', port=8545): Read timed out. (read timeout=10)
+            # from Go Ethereum. This translates to the error "context was cancelled" on the server side:
+            # https://github.com/ethereum/go-ethereum/issues/20426
+            if i < retries - 1:
+                # Give some more verbose info than the default middleware
+                logger.warning(
+                    "Retrying events for block range %d - %d (%d) failed with %s, retrying in %s seconds",
+                    start_block,
+                    end_block,
+                    end_block - start_block,
+                    e,
+                    delay,
+                )
+                # Decrease the `eth_get_blocks` range
+                end_block = start_block + ((end_block - start_block) // 2)
+                # Let the JSON-RPC to recover e.g. from restart
+                time.sleep(delay)
+                continue
+            else:
+                logger.warning("Out of retries")
+                raise
+    return None
```

Comment:
Would this ever return None - it should either return the end_block. return of func or raise an exception?

Reply:
True. The None was to satisfy the linter which was somehow unhappy